File: lab-redes/python/tcp/cliente_tcp.py
import os
import socket
import sys

HOST = "localhost"
PORTA_BASE = 5000


def ler_offset() -> int:
    if len(sys.argv) > 1 and sys.argv[1].isdigit():
        return int(sys.argv[1])
    return int(os.getenv("LAB_REDES_OFFSET", "0"))


def mensagens_script() -> list[str]:
    inicio = 2 if len(sys.argv) > 1 and sys.argv[1].isdigit() else 1
    return sys.argv[inicio:]


def enviar(cliente: socket.socket, arquivo, mensagem: str) -> None:
    print(f"> {mensagem}", flush=True)
    cliente.sendall((mensagem + "\n").encode("utf-8"))
    print(arquivo.readline().strip(), flush=True)


porta = PORTA_BASE + ler_offset()
mensagens = mensagens_script()

with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as cliente:
    cliente.connect((HOST, porta))
    print(f"[TCP] Conectado ao servidor em {HOST}:{porta}.", flush=True)
    arquivo = cliente.makefile("r", encoding="utf-8")

    if mensagens:
        for mensagem in mensagens:
            enviar(cliente, arquivo, mensagem)
            if mensagem.lower() == "sair":
                break
    else:
        print("[TCP] Digite 'sair' para encerrar.", flush=True)
        while True:
            mensagem = input("> ")
            enviar(cliente, arquivo, mensagem)
            if mensagem.lower() == "sair":
                break

File: lab-redes/java/tcp/ClienteTCP.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.Socket;
import java.util.Arrays;
import java.util.List;

public class ClienteTCP {
    private static final int PORTA_BASE = 5000;

    public static void main(String[] args) throws IOException {
        String host = "localhost";
        int offset = lerOffset(args);
        int porta = PORTA_BASE + offset;
        List<String> mensagensScript = mensagensScript(args);

        try (Socket socket = new Socket(host, porta);
             PrintWriter saida = new PrintWriter(socket.getOutputStream(), true);
             BufferedReader entrada = new BufferedReader(new InputStreamReader(socket.getInputStream()));
             BufferedReader teclado = new BufferedReader(new InputStreamReader(System.in))) {

            System.out.println("[TCP] Conectado ao servidor em " + host + ":" + porta + ".");

            if (!mensagensScript.isEmpty()) {
                for (String mensagem : mensagensScript) {
                    enviarMensagem(saida, entrada, mensagem);
                    if (mensagem.equalsIgnoreCase("sair")) {
                        break;
                    }
                }
                return;
            }

            System.out.println("[TCP] Digite 'sair' para encerrar.");
            while (true) {
                System.out.print("> ");
                String linha = teclado.readLine();
                enviarMensagem(saida, entrada, linha);
                if (linha.equalsIgnoreCase("sair")) {
                    break;
                }
            }
        }
    }

    private static void enviarMensagem(PrintWriter saida, BufferedReader entrada, String mensagem) throws IOException {
        System.out.println("> " + mensagem);
        saida.println(mensagem);
        System.out.println(entrada.readLine());
    }

    private static int lerOffset(String[] args) {
        if (args.length > 0 && args[0].matches("\\d+")) {
            return Integer.parseInt(args[0]);
        }

        String valor = System.getenv("LAB_REDES_OFFSET");
        return valor == null || valor.isBlank() ? 0 : Integer.parseInt(valor);
    }

    private static List<String> mensagensScript(String[] args) {
        int inicio = args.length > 0 && args[0].matches("\\d+") ? 1 : 0;
        return Arrays.asList(args).subList(inicio, args.length);
    }
}
